### PR TITLE
feat: support for extra pre-stop hook command

### PR DESCRIPTION
**Pull request by zxh326**
doc updated later

### CODE REVIEW

File: pkg/config/setting.go [hunk 1]
```diff
@@ -297,6 +298,10 @@ func ParseSetting(secrets, volCtx map[string]string, options []string, usePod bo
 		jfsSetting.Envs = env
 	}
 
+	if secrets["extraPreStopHookCmd"] != "" {
+		jfsSetting.ExtraPreStopHookCmd = secrets["extraPreStopHookCmd"]
```
Comment (timfeirg) [inline pkg/config/setting.go]: can we simply use `prestop`?

the thing is ... we already have 2 types of naming conventions in this json struct (`secret-key|envs_map`), can't help to feel reluctant to add a third `extraPreStopHookCmd`

File: pkg/juicefs/mount/builder/common.go [hunk 2]
```diff
@@ -96,10 +96,14 @@ func (r *BaseBuilder) genCommonJuicePod(cnGen func() corev1.Container) *corev1.P
 		},
 	}}
 	pod.Spec.Containers[0].Resources = r.jfsSetting.Resources
+	preStopHookCmd := fmt.Sprintf(
+		"umount %s -l; rmdir %s; exit 0", r.jfsSetting.MountPath, r.jfsSetting.MountPath)
+	if r.jfsSetting.ExtraPreStopHookCmd != "" {
+		preStopHookCmd = fmt.Sprintf("%s; %s", r.jfsSetting.ExtraPreStopHookCmd, preStopHookCmd)
```
Comment (timfeirg) [inline pkg/juicefs/mount/builder/common.go:102]: strip the trailing `;` from `r.jfsSetting.ExtraPreStopHookCmd`, if user happens to write something like `echo 1;`, the resulting double `;;` will cause error

File: pkg/config/setting.go (hunk 1 above)
Comment (zxh326) [reply pkg/config/setting.go]: what about `extra-prestop-cmd`

user needs to be know that this is an **extra** cmd and not an overwrite?

File: pkg/juicefs/mount/builder/pod_test.go [hunk 3]
```diff
@@ -356,6 +361,16 @@ func TestNewMountPod(t *testing.T) {
 			},
 			want: podMetricTest,
 		},
+		{
+			name: "test-extre-pre-stop-hook-cmd",
+			args: args{
+				name:                "test",
+				cmd:                 defaultCmd,
+				mountPath:           defaultMountPath,
+				extraPreStopHookCmd: "sleep 30",
```
Comment (timfeirg) [inline pkg/juicefs/mount/builder/pod_test.go]: consider using somewhat complex example like `sh -c "cd /tmp && ls; echo 1"`

File: pkg/config/setting.go (hunk 1 above)
Comment (timfeirg) [reply pkg/config/setting.go]: > user needs to be know that this is an extra cmd and not an overwrite?

this is a bit like helm values, where all fields are in fact "extra", but they do not have the "extra" word as prefix, like `annotations`, `labels`

we'll introduce the minor detail in the docs, but even if users misunderstood and treat it like an overwrite, it won't hurt anything, correct?

File: pkg/config/setting.go [hunk 4]
```diff
@@ -56,14 +56,15 @@ type JfsSetting struct {
...
-	Configs       map[string]string `json:"configs_map,omitempty"`
+	SecretKey           string            `json:"secret-key,omitempty"`
+	SecretKey2          string            `json:"secret-key2,omitempty"`
+	Token               string            `json:"token,omitempty"`
+	Passphrase          string            `json:"passphrase,omitempty"`
+	Envs                map[string]string `json:"envs_map,omitempty"`
+	EncryptRsaKey       string            `json:"encrypt_rsa_key,omitempty"`
+	InitConfig          string            `json:"initconfig,omitempty"`
+	Configs             map[string]string `json:"configs_map,omitempty"`
+	ExtraPreStopHookCmd string            `json:"-,omitempty"`
```
Comment (timfeirg) [inline pkg/config/setting.go:67]: field name `prestop-cmd`?

**Review by timfeirg — APPROVED** [review]

**Pull request closed without merging**